FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1892 @@
+  // and the actual FPU stack.
+  Ref SynchronizeStackValues();
+  // Moves us from the fast to the slow path if ShouldMigrate is true.
+  void MigrateToSlowPathIf(bool ShouldMigrate);
+  // Top Cache Management
+  Ref GetTopWithCache_Slow();
+  Ref GetOffsetTopWithCache_Slow(uint8_t Offset);
+  void SetTopWithCache_Slow(Ref Value);
+  Ref GetX87ValidTag_Slow(uint8_t Offset);
+  // Resets fields to initial values
+  void Reset(bool AlsoSlowPath = true);
+
+  struct StackMemberInfo {
+    IR::OpSize SourceDataSize; // Size of SourceDataNode
+    IR::OpSize StackDataSize;  // Size of the data in the stack
+    IR::Ref SourceDataNode;    // Reference to the source location of the data.
+                               // In the case of a load, this is the source node of the load.
+                               // If it's not a load, then it's nullptr.
+    IR::Ref StackDataNode;     // Reference to the data in the Stack.
+    bool InterpretAsFloat {};  // True if this is a floating point value, false if integer
+  };
+
+  // StackData, TopCache need to be always properly set to ensure
+  // it reflects the current state of the FPU. This sync only makes sense while
+  // taking the fast path. Once in the slow path, these don't make sense anymore
+  // and we are syncing everything.
+
+  // Index on vector is offset to top value at start of block
+  // If slow path is true, then StackData is always empty.
+  FixedSizeStack<StackMemberInfo, 8> StackData;
+  using StackSlot = FixedSizeStack<StackMemberInfo, 8>::Slot;
+
+  void InvalidateCaches();
+  void InvalidateTopOffsetCache();
+
+  // Cache for Constants
+  // ConstantPoll[i] has IREmit->_Constant(i);
+  constexpr static size_t ConstantPoolSize = 8;
+  std::array<Ref, ConstantPoolSize> ConstantPool;
+  Ref GetConstant(ssize_t Offset);
+
+  // Cached value for Top
+  // If slowpath is false, then TopCache is nullptr.
+  std::array<Ref, 8> TopOffsetCache;
+  // Are we on the slow path?
+  // Once we enter the slow path, we never come out.
+  // This just simplifies the code atm. If there's a need to return to the fast path in the future
+  // we can implement that but I would expect that there would be very few cases where that's necessary.
+  // On the slow path TopCache is always the last obtained version of top.
+  // TopOffset is ignored
+  bool SlowPath = false;
+  // Keeping IREmitter not to pass arguments around
+  IREmitter* IREmit = nullptr;
+};
+
+inline void X87StackOptimization::InvalidateCaches() {
+  LogMan::Msg::DFmt("Invalidating caches");
+
+  InvalidateTopOffsetCache();
+  ConstantPool.fill(nullptr);
+}
+
+inline void X87StackOptimization::InvalidateTopOffsetCache() {
+  TopOffsetCache.fill(nullptr);
+}
+
+inline void X87StackOptimization::Reset(bool AlsoSlowPath) {
+  if (AlsoSlowPath) {
+    SlowPath = false;
+  }
+  StackData.clear();
+  InvalidateCaches();
+}
+
+inline Ref X87StackOptimization::GetConstant(ssize_t Offset) {
```

Comment:
We already have a general-purpose constant pooling optimization in ConstProp, this is strictly a specific case of it. Does it make sense to have both? (for example to improve compile-time?)

Reply:
Not really - we could potentially abstract into a ConstantPool class, what do you think?

Reply:
No, the logic is different anyway. I don't mean that "we shouldn't have 2 separate implementations of constant pooling logic". Rather that "we're going to have constant pooling happen anyway if the pass order is right, so is it worth also pooling constants here ahead of time?"

(The answer may well be yes.)

Reply:
TBH, with a view to future IR work, I think I'd weakly prefer if we drop the pooling and just use `_Constant` throughout the file to remain consistent with the OpcodeDispatcher & the rest of FEX. This feels like a premature optimization.